File: sample_code/cplusplus/Tetris/NextWidget.cpp

#include "NextWidget.h"

void NextWidget::paintEvent(QPaintEvent * e) {
    QFrame::paintEvent(e);
    const QColor orange(255, 127, 0);
    static const QColor colorTable[] = {Qt::black, Qt::magenta, Qt::green, Qt::cyan, orange, Qt::blue, Qt::red, Qt::yellow};
    int t = 2 * lineWidth() + midLineWidth(), *c, i, x, k;
    QPainter painter(this);

    int __action__ = 4;

    while(true) {
        switch(__action__) {
            case 0:
            if(nextBlock == 0) {
                __action__ = 7;
            }
            else {
                __action__ = 5;
            }
            break;

            case 1:
            if(i < 8) {
                __action__ = 3;
            }
            else {
                __action__ = 6;
            }
            break;

            case 3:
            painter.fillRect((c[i] - 3 + x) * 20 + t + 1, c[i + 1] * 20 + t + 1, 18, 18, colorTable[k]);
            i += 2;
            __action__ = 1;
            break;

            case 4:
            painter.fillRect(t, t, 120, 80, colorTable[0]);
            __action__ = 0;
            break;

            case 5:
            c = nextBlock->getCoordinates();
            k = nextBlock->getKind();
            x = k == Block::I || k == Block::O;
            i = 0;
            __action__ = 1;
            break;

            case 6:
            return;
            case 7:
            return;
        }
    }
}

void NextWidget::draw(Block * b) {
    nextBlock = b;
    update();
}

NextWidget::NextWidget() {
    int border;

    nextBlock = 0;

    setFrameStyle(QFrame::Box | QFrame::Raised);
    setLineWidth(4);
    setMidLineWidth(3);

    border = 4 * lineWidth() + 2 * midLineWidth();
    setFixedSize(120 + border, 80 + border);
}

File: sample_code/java/Tetris/NextWidget.java

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import javax.swing.JPanel;

public class NextWidget extends JPanel {
    private Block nextBlock;

    public void paint(Graphics g) {
        Color [] colorTable = {Color.black, Color.magenta, Color.green, Color.cyan, Color.orange, Color.blue, Color.red, Color.yellow}; 
        int x = 0, k = 0, i = 0;
        int [] c = null;

        int __action__ = 3;

        __loop__: while(true) {
            switch(__action__) {
                case 0:
                if(nextBlock == null) {
                    __action__ = 6;
                }
                else {
                    __action__ = 4;
                }
                break;

                case 2:
                if(i < 8) {
                    __action__ = 5;
                }
                else {
                    __action__ = 6;
                }
                break;

                case 3:
                g.setColor(Color.black);
                g.fillRect(0, 0, 120, 80);

                __action__ = 0;
                break;

                case 4:
                c = nextBlock.getCoordinates();
                k = nextBlock.getKind();
                x = k == Block.I || k == Block.O ? 1 : 0;
                i = 0;
                __action__ = 2;
                break;

                case 5:
                g.setColor(colorTable[nextBlock.getKind()]);
                g.fillRect((c[i] - 3 + x) * 20 + 1, c[i + 1] * 20 + 1, 18, 18);
                i += 2;
                __action__ = 2;
                break;

                case 6:
                break __loop__;
            }
        }
    }

    public void draw(Block b) {
        nextBlock = b;
        repaint();
    }

    public NextWidget() {
        int border;

        nextBlock = null;

        setPreferredSize(new Dimension(120, 80));
    }
}
